lib/route-helpers.js

```diff
@@ -0,0 +1,229 @@
+  return _.isArray( fieldSchema ) &&
+    _.isObject( fieldSchema[ 0 ] ) &&
+    !_.has( fieldSchema[ 0 ], 'ref' );
+};
+
+exports.processReplaceOp = function(op, model){
+  var field = op.path.split('/').slice(3);
+  var value = op.value;
+  var result = {
+    match: {},
+    separate: false,
+    key: '$set',
+    update: {}
+  };
+  var ret = [result];
+
+  if (field.length > 1 && field[0] !== "links") {
+    var needsPositionalUpdate = exports.needsPositionalUpdate(field, model);
+
+    if (needsPositionalUpdate){
+      //happy to assume it's working one-level only for now
+      var subdocid = field[1];
+      var embeddedPath = field[0];
+      result.match[embeddedPath + '._id'] = subdocid; //new ObjectId(subdocid);
+      result.separate = true;
+      var updatePath = [embeddedPath, '$'].concat(field.slice(2)).join('.');
+      result.update[updatePath] = value;
+
+      //And add an op updating the possible deleted bit
+      var forDeleted = {match: {}, separate: true, key: '$set', update: {}};
+      forDeleted.match['_internal.deleted.'+embeddedPath + '._id'] = subdocid; //new ObjectId(subdocid);
+      forDeleted.update['_internal.deleted.'+ updatePath] = value;
+      ret.push(forDeleted);
+    }else{
+      //regular update to deep path
+      result.update[field.join(".")] = value;
+    }
+  } else {
+    result.update[field[field.length-1]] = value;
+  }
+  return ret;
+};
+
+exports.processAddOp = function(op, model){
+  var result = {
+    match: {},
+    separate: false,
+    key: '$pushAll',
+    update: {}
+  };
+
+  var field = op.path.split('/').splice(3);
+
+  var tabFactor = field[field.length - 1] === "-" ? 2 : 1;
+  if (field.length > 1 && field[0] !== "links") {
+    var needsPositionalUpdate = exports.needsPositionalUpdate(field, model);
+
+    if (needsPositionalUpdate && (field.length - tabFactor) > 1){
+      var subdocid = field[1];
+      var embeddedPath = field[0];
+      result.match[embeddedPath + '._id'] = subdocid;
+      result.separate = true;
+      var updatePath = [embeddedPath, '$'].concat(
+        tabFactor === 2 ? _.initial(field, field.length - 1).slice(2) : field.slice(2)
+      ).join('.');
+      result.update[updatePath] = result.update[updatePath] || [];
+      result.update[updatePath].push(op.value);
+    }else{
+      var normalisedPath = (tabFactor === 2 ? _.initial(field, field.length - 1) : field).join('.');
+
+      result.update[normalisedPath] = result.update[normalisedPath] || [];
+      result.update[normalisedPath].push(op.value);
+    }
+  } else{
+    result.update[field[field.length - tabFactor]] = result.update[field[field.length - tabFactor]] || [];
+    result.update[field[field.length - tabFactor]].push(op.value);
+  }
+
+  return [result];
+};
+
+exports.processRemoveOp = function(op, model, returnedDocument){
+  var field = op.path.split('/').slice(3);
+  var value = field.pop()
+    , fieldName, pullKey, fieldSchema, nestedObjects;
+
+  if( field.length === 1 ){
```

Comment:
Fields like this block is very similar to `needsPositionalUpdate` function. Any way to refactor it and reuse it here?